ion-schema/src/model/ranges.rs

```diff
@@ -0,0 +1,347 @@
+            };
+            list_writer.close()?;
+        }
+        Ok(())
+    }
+}
+
+const MIN: &str = "min";
+const MAX: &str = "max";
+const EXCLUSIVE: &str = "exclusive";
+const RANGE: &str = "range";
+
+impl<V: IslVersion, T: RangeBoundRequirements + ReadFromIsl<V>> ReadFromIsl<V>
+    for IonSchemaRange<T>
+{
+    fn try_read(ion: &Element, ctx: &LoaderContext<V>) -> IonSchemaResult<Self> {
+        let optional_annotation = ion.one_optional_annotation()?;
+        if optional_annotation == Some(RANGE) {
+            let Some(seq) = ion.as_list() else {
+                invalid_schema_error(format!("range must be a non-null list; found: {ion}"))?
+            };
+
+            isl_require!(seq.len() == 2 => "range must have a lower and upper bound; found: {ion}")?;
+            // We can unwrap these without panicking
+            let lower = seq.get(0).unwrap();
+            let upper = seq.get(1).unwrap();
+
+            let lower = match (lower.as_symbol_text(), lower.one_optional_annotation()?) {
+                (Some(MIN), Some(EXCLUSIVE)) => invalid_schema_error("'min' may not be exclusive")?,
+                (Some(MIN), None) => Unbounded,
+                (_, Some(EXCLUSIVE)) => Excluded(T::try_read(lower, ctx)?),
+                (_, None) => Included(T::try_read(lower, ctx)?),
+                _ => invalid_schema_error(format!("invalid annotation on range bound: {lower}"))?,
+            };
+            let upper = match (upper.as_symbol_text(), upper.one_optional_annotation()?) {
+                (Some(MAX), Some(EXCLUSIVE)) => invalid_schema_error("'max' may not be exclusive")?,
+                (Some(MAX), None) => Unbounded,
+                (_, Some(EXCLUSIVE)) => Excluded(T::try_read(upper, ctx)?),
+                (_, None) => Included(T::try_read(upper, ctx)?),
+                _ => invalid_schema_error(format!("invalid annotation on range bound: {upper}"))?,
+            };
+
+            Ok(IonSchemaRange::try_new(lower, upper)?)
+        } else {
+            isl_require!(optional_annotation.is_none() => "invalid annotation on range; found: {ion}")?;
+            let value: T = T::try_read(ion, ctx)
+                .map_err(|_| invalid_schema_error_raw(format!("invalid range value: {ion}")))?;
+            Ok(IonSchemaRange::from(value))
+        }
+    }
+}
+
+impl<T: RangeBoundRequirements> From<&T> for IonSchemaRange<T> {
+    fn from(value: &T) -> Self {
+        IonSchemaRange {
+            min: Included(value.clone()),
+            max: Included(value.clone()),
+        }
+    }
+}
+impl<T: RangeBoundRequirements> From<T> for IonSchemaRange<T> {
+    fn from(value: T) -> Self {
+        IonSchemaRange {
+            min: Included(value.clone()),
+            max: Included(value),
+        }
+    }
+}
+
+// Technically, the implementations this produces can construct empty ranges, which are not valid in
+// Ion Schema Language. This only affects programmatically constructed types. If we want to
```

Comment:
Given our re-read of the spec, should this be updated?

Reply:
I've created https://github.com/amazon-ion/ion-schema/issues/140. Once that is resolved, I think we can remove this note. (I'll also update this note to mention that issue.)